From 724b86883923b4b5ad84ce38c0a2efdaef871e50 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Wed, 30 Apr 2025 13:44:22 +0800
Subject: [PATCH 177/254] mtk: hostapd: fix start_disabled

Current start_disabled would not send beacon and set channdel to driver.
So there is no bss_conf in mt76 and lead to setup interface fail.
In this patch, set beacon and set channel as normal flow and disable
beacon by vendor command when start_disabled is set.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 src/ap/hostapd.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 222d0f876..bbe37bd74 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1369,7 +1369,7 @@ static int hostapd_start_beacon(struct hostapd_data *hapd,
 {
 	struct hostapd_bss_config *conf = hapd->conf;
 
-	if (!conf->start_disabled && ieee802_11_set_beacon(hapd) < 0)
+	if (ieee802_11_set_beacon(hapd) < 0)
 		return -1;
 
 #ifdef CONFIG_IEEE80211BE
@@ -1385,7 +1385,7 @@ static int hostapd_start_beacon(struct hostapd_data *hapd,
 		}
 	}
 
-	if (flush_old_stations && !conf->start_disabled &&
+	if (flush_old_stations &&
 	    conf->broadcast_deauth && (hapd->conf->mld_ap && !hapd->mld->started)) {
 		u8 addr[ETH_ALEN];
 
@@ -1407,6 +1407,9 @@ static int hostapd_start_beacon(struct hostapd_data *hapd,
 	hostapd_ubus_add_bss(hapd);
 	hostapd_ucode_add_bss(hapd);
 
+	if (conf->start_disabled)
+		hostapd_drv_beacon_ctrl(hapd, false);
+
 	return 0;
 }
 
-- 
2.45.2

